rfcs/text/1020-link-entities.md

```diff
@@ -0,0 +1,385 @@
+
+# Motivation
+
+[motivation]: #motivation
+
+## Preface: Edges within the graph
+
+At the moment, the Type System is a fairly 'complete' system.
+The primitives defined within it are capable of being composed together to model highly complex graph topologies, and arbitrary JSON within an entity's properties.
+
+Such representations are encoded through references between elements, an entity type referring to a property type within its schema for example.
+These inter-element references are the glue that makes up the graph and as such will be referred to as "edges".
+
+> **Note**: The term "edge" is distinct from "link", it is used here to refer to the relationship between two graph elements which include elements of the type system _as well as_ entities.
+
+At the present, the following types of edges exist:
+
+- Entity Type -> Link Type (constrains the types of links that can come from an entity)
+- Entity Type -> Entity Type (constrains the types of entities that can be linked to from an entity)
+- Entity Type -> Property Type (constrains the properties that can appear on an entity)
+- Property Type -> Property Type (constrains the types of properties that can appear within an object in another property)
+- Property Type -> Data Type (constrains the type of data that can be stored within a property)
+- Entity -> Entity Type (describes the associated entity type of an entity)
+- Entity -> Link (the outgoing or incoming link of an entity)
+- Link -> Entity (the source or target entity of a link)
+
+## Creating Higher-order abstractions through composition of the primitives of the type system
+
+### Groups
+
+Due to the compositional design of the primitives of the Type System, it's possible to create high-order abstractions within the system.
+For example, one could want to define "groups" of entities within their system, in such a way that they would be programmatically resolvable through blocks, _with constraints guaranteed by the specification_.
+
+One could currently do this by creating a "Group" entity type, and an associated "Has Element" link type.
+They could then create a specific "Group" entity, and link it to the entities they wish to be part of the group:
+
+```text
+       ┌─────┐
+       │Group│
+       └──┬──┘
+          │
+          │
+ ┌───Has element───┐
+ │     │     │     │
+ │     │     │     │
+ ▼     ▼     ▼     ▼
+┌─┐   ┌─┐   ┌─┐   ┌─┐
+│A│   │B│   │C│   │D│
+└─┘   └─┘   └─┘   └─┘
+```
+
+Being able to create such abstractions is a powerful feature of the Type System, as users are able to implicitly benefit from some of the following scenarios:
+
+- A user updates entity `A`, the link is still associated with the new version
+- A user deletes entity `A`, the Embedding Application (EA) knows that a link has been invalidated and is able to orchestrate/trigger a resolution flow
+- A user redefines the destination constraints of `Has element` links on the `Group` entity type to disallow entities of type `D`.
```

Comment:
Explaining them feels a bit superfluous but happy to do so if you think it's confusing to not have, or otherwise helpful to do so.

- The first one is implicit in the Block Protocol rather than the type system, updating an entity does not change its `EntityId`, identity is preserved, and links are based on identity.
- Similarly, with the second one, the application is _explicitly required to be aware of those primitives_, I do not believe "I think this applies to most if not all ways of modelling relational data" is quite true in this case. The point is that if users start creating their own relationship definitions, e.g. a block decides to just store an `EntityId` in a property, the application is not _explicitly_ aware of that link, and doesn't know it's invalidated. Sure, it could _theoretically_ be introspecting every single property ever and trying to intelligently identify relationships, but that's not an expectation of EAs.
- Your explanation of what it's due to is correct. As that's not actually a _part_ of this RFC, I do think it perhaps doesn't need to be fully explained as it's captured as a core behaviour of the system and explaining it serves to extend the RFC with calling out to other features.. 

I'll open a new comment with a suggestion because unfortunately the range in this one is missing the last line

Reply:
Does this help:
[Add some supporting information to example situations](https://github.com/blockprotocol/blockprotocol/pull/1020/commits/0fe9b290b2ae5c55bf75fd691865e12e4de2db0c)

Reply:
Yes, thank you